cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/BrokerSetAwareGoal.java

```diff
@@ -0,0 +1,380 @@
+        GoalUtils.ensureNoOfflineReplicas(clusterModel, name());
+        // Sanity check: No replica should be moved to a broker, which used to host any replica of the same partition on its broken disk.
+        GoalUtils.ensureReplicasMoveOffBrokersWithBadDisks(clusterModel, name());
+        // Sanity check to confirm that the final distribution is broker set aware.
+        ensureBrokerSetAware(clusterModel, optimizationOptions);
+        if (_provisionResponse.status() != ProvisionStatus.OVER_PROVISIONED) {
+            _provisionResponse = new ProvisionResponse(ProvisionStatus.RIGHT_SIZED);
+        }
+        finish();
+    }
+
+    private void ensureBrokerSetAware(ClusterModel clusterModel, OptimizationOptions optimizationOptions)
+            throws OptimizationFailureException {
+        // Sanity check to confirm that the final distribution is brokerSet aware.
+        Set<String> excludedTopics = optimizationOptions.excludedTopics();
+        Map<String, Set<Broker>> brokersByBrokerSet = _brokerSetDataHelper.getBrokersByBrokerSet();
+
+        // Sanity check to check if brokersByBrokerSet map is not empty
+        if (brokersByBrokerSet == null || brokersByBrokerSet.isEmpty()) {
+            throw new OptimizationFailureException(String.format("[%s] broker set to brokers mapping cannot be empty.",
+                                                                 name()));
+        }
+
+        // Sanity check to check if all brokers in data store do not match all brokers in cluster model
+        Set<Broker> allBrokersInDataStore = brokersByBrokerSet.values()
+                .stream()
+                .flatMap(brokers -> brokers.stream())
+                .collect(Collectors.toSet());
+
+        if (!allBrokersInDataStore.equals(clusterModel.brokers())) {
+            throw new OptimizationFailureException(String.format("[%s] All Brokers from data store %s do not match brokers in cluster model %s.",
+                                                                 name(), allBrokersInDataStore, clusterModel.brokers()));
+        }
+
+        for (Map.Entry<String, List<Partition>> partitionsByTopic : clusterModel.getPartitionsByTopic()
+                .entrySet()) {
+            if (!excludedTopics.contains(partitionsByTopic.getKey())) {
+                List<Partition> partitions = partitionsByTopic.getValue();
+                Set<Broker> allBrokersForTopic = partitions.stream()
+                        .map(partition -> partition.partitionBrokers())
+                        .flatMap(brokers -> brokers.stream())
+                        .collect(Collectors.toSet());
+                // Checks if a partition's brokers do not all live in a single brokerSet
+                if (!brokersByBrokerSet.values()
+                        .stream()
+                        .anyMatch(brokerSetBrokers -> brokerSetBrokers.containsAll(allBrokersForTopic))) {
```

Comment:
Nit `'!Stream.anyMatch(...)' can be replaced with 'noneMatch()'`

Reply:
Updated